refactor: Remove Span operator==, Use std::ranges::equal

maflcko opened the pull request:
`std::span` removed the comparison operators, so it makes sense to remove them for the `Span` "backport" as well. Using `std::ranges::equal` also has the benefit that some `Span` temporary constructions can now be dropped.

This is required to move from `Span` toward `std::span`.

Commit:
refactor: Remove Span operator==, Use std::ranges::equal

fanquake commented:
> Looks like this also fails on macOS on GHA.

Pretty sure that job should be rerun, because it's failing in the brew install stage.

fanquake commented:
> Does anyone know what version(s) of clang can be expected to be available on macOS normally?

GIven our minimum supported macOS (11.x), you could run anything from Xcode 12.5+ onwards, which is roughly LLVM/Clang 12+. Although I'd expect most users compiling from source on macOS, would be using a much more recent Xcode, so likely Clang ~14+ at least.

Also, anyone should be able to install and run the latest LLVM/Clang via brew.

vasild commented:
Shouldn't our `Span` be dropped completely, now that we have C++20 and `std::span`?

vasild commented:
Concept ACK

hodlinator commented on src/test/miniscript_tests.cpp:
```diff
@@ -22,6 +18,12 @@
 #include <script/script_error.h>
 #include <script/signingprovider.h>
 
+
```
nit: Double newline

hodlinator commented on src/wallet/db.cpp:
```diff
@@ -9,15 +9,16 @@
 namespace wallet {
-bool operator<(BytePrefix a, Span<const std::byte> b) { return a.prefix < b.subspan(0, std::min(a.prefix.size(), b.size())); }
-bool operator<(Span<const std::byte> a, BytePrefix b) { return a.subspan(0, std::min(a.size(), b.prefix.size())) < b.prefix; }
+bool operator<(BytePrefix a, Span<const std::byte> b) { return std::ranges::lexicographical_compare(a.prefix, b.subspan(0, std::min(a.prefix.size(), b.size()))); }
+bool operator<(Span<const std::byte> a, BytePrefix b) { return std::ranges::lexicographical_compare(a.subspan(0, std::min(a.size(), b.prefix.size())), b.prefix); }
```
(Attempted switching around the parameters to `lexicographical_compare()` and confirmed that it broke `test_bitcoin -t db_tests`).

hodlinator reviewed (approved):
ACK fabcd0c4fe44e3bc2d6a59f2839f459fd5c81171

hodlinator reviewed (approved):
Untested re-ACK fa69fba751079ee6042daceb88bc8f8ad3f3de96

`git range-diff master fabcd0c fa69fba` reveals pure whitespace change.

stickies-v reviewed (commented):
Approach ACK

Commit:
refactor: Use std::ranges::equal in GetNetworkForMagic

Replace std::equal with std::ranges::equal, because it allows for
shorter code, because no pointers or iterators have to be passed
explicitly.

hodlinator reviewed (approved):
Untested Code Review re-ACK fad0cf6

`git range-diff master fa69fba fad0cf6` on shows `#include` changes and new fad0cf6f26 commit.

`git show fad0cf6` shows use of `std::equal` overload with 3 iterators being replaced with the *generally* safer `std::ranges::equal` (not really an issue in this specific case as we are comparing pairs of the same type `MessageStartChars = std::array<uint8_t, 4>`).

stickies-v commented on src/test/fuzz/poolresource.cpp:
```diff
@@ -78,7 +78,7 @@ class PoolResourceFuzzer
         std::vector<std::byte> expect(entry.span.size());
         InsecureRandomContext(entry.seed).fillrand(expect);
-        assert(entry.span == expect);
+        assert(std::ranges::equal(entry.span, expect));
```
nit: missing `#include <algorithm>`

stickies-v commented on src/kernel/chainparams.cpp:
```diff
@@ -689,15 +689,15 @@ std::optional<ChainType> GetNetworkForMagic(const MessageStartChars& message)
     const auto signet_msg = CChainParams::SigNet({})->MessageStart();
 
-    if (std::equal(message.begin(), message.end(), mainnet_msg.data())) {
+    if (std::ranges::equal(message, mainnet_msg)) {
```
nit: I think we can use `std::array::operator==` here? (edit: or a switch statement would probably be even better)

```suggestion
    if (message == mainnet_msg) {
```

<details>
<summary>git diff on fad0cf6f26</summary>

```diff
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index 7d06d24be5..a760926ec8 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -689,15 +689,15 @@ std::optional<ChainType> GetNetworkForMagic(const MessageStartChars& message)
     const auto regtest_msg = CChainParams::RegTest({})->MessageStart();
     const auto signet_msg = CChainParams::SigNet({})->MessageStart();
 
-    if (std::ranges::equal(message, mainnet_msg)) {
+    if (message == mainnet_msg) {
         return ChainType::MAIN;
-    } else if (std::ranges::equal(message, testnet_msg)) {
+    } else if (message == testnet_msg) {
         return ChainType::TESTNET;
-    } else if (std::ranges::equal(message, testnet4_msg)) {
+    } else if (message == testnet4_msg) {
         return ChainType::TESTNET4;
-    } else if (std::ranges::equal(message, regtest_msg)) {
+    } else if (message == regtest_msg) {
         return ChainType::REGTEST;
-    } else if (std::ranges::equal(message, signet_msg)) {
+    } else if (message == signet_msg) {
         return ChainType::SIGNET;
     }
     return std::nullopt;

```
</details>

stickies-v replied:
That makes sense, but I don't quite see the case for changing `GetNetworkForMagic` to accept a span in the future, so it feels a bit like premature optimization to me, but I may be missing context.

It's a nit, and the commit is orthogonal to the PR anyway, so this is the last I'll comment on it - up to you what you prefer, but I feel like this would probably be the most elegant implementation:

```cpp
std::optional<ChainType> GetNetworkForMagic(const MessageStartChars& message)
{
    static const std::array<std::pair<MessageStartChars, ChainType>, 5> network_map = {{
        {CChainParams::Main()->MessageStart(), ChainType::MAIN},
        {CChainParams::TestNet()->MessageStart(), ChainType::TESTNET},
        {CChainParams::TestNet4()->MessageStart(), ChainType::TESTNET4},
        {CChainParams::RegTest({})->MessageStart(), ChainType::REGTEST},
        {CChainParams::SigNet({})->MessageStart(), ChainType::SIGNET}
    }};

    auto it = std::ranges::find_if(network_map, [&message](const auto& pair) { return pair.first == message; });
    return (it != network_map.end()) ? std::make_optional(it->second) : std::nullopt;
}
```

stickies-v reviewed (approved):
ACK fad0cf6f2619df8df435a2da6da49eeb5510a10f

Even though not having span comparison operators can make things a bit more verbose, these changes are needed to start using `std::span` over our own implementation. This approach seems like the least invasive one (cleaning up wherever possible), and I can't see any behaviour change.

sedited reviewed (approved):
ACK fad0cf6f2619df8df435a2da6da49eeb5510a10f

The last commit seems a bit random, why change that function and not any of the other occurrences of `std::equal`?
